src/dbnode/persist/fs/retriever.go

```diff
@@ -946,19 +948,18 @@ func (req *retrieveRequest) Segment() (ts.Segment, error) {
 func (req *retrieveRequest) Finalize() {
 	// May not actually finalize the request, depending on if
 	// retriever is done too
-	if req.finalized {
+	if req.finalized.Swap(true) {
```

Comment:
I think you're changing the semantics significantly here, not sure if intentionally?
Now it's *not* possible to call `onCallerOrRetrieverDone` multiple times, so the done fn will not actually release the request back to the pool, as it only does that when called twice: once by caller, once by retriever.
Maybe `onCallerOrRetrieverDone` should be renamed to `onCallerAndRetrieverDone`?

Think it makes more sense to remove finalizes *accounting* from `onCallerOrRetrieverDone` and just do:
```
func (req *retrieveRequest) Finalize() {
	if req.finalized.Load() >= 2 {
		return
	}

	// Wait until request is complete before attempting to finalize
	req.resultWg.Wait()

	// May not actually finalize the request, depending on if
	// retriever is done too
	if req.finalized.Inc() == 2 {
		req.onCallerAndRetrieverDone()
	}
}
```